Iqss/6485 - Multiple File Stores

qqmyers opened the pull request:
**What this PR does / why we need it**: Implements the ability to configure multiple file stores.

**Which issue(s) this PR closes**:

Closes #6485 

**Special notes for your reviewer**: The install script should set up a file store, backward compatible with an existing installation. The docs also describe how to do this.

**Suggestions on how to test this**: Multiple things that could be tested:
Basic install comes up with a single file store working as before (can upload/download files)
Ability to configure a second file store (e.g. with a different file path) 
Ability to configure an additional s3 store 
Ability to configure another s3 store using a separate aws profile
Ability to change the default store for new files (new store allows upload/download, files in other stores are still downloadable)
Ability to list stores available and use the admin API and/or superuser UI (both described in docs) to set/reset to default which store is used for a given dataverse.
Verify that a store set for a given dataverse works for datasets created/edited within that dataverse.
With a s3 store configured with the id 's3', verify that rsync still works
Without an s3 store configured with the id 's3', verify that rsync upload does not appear as an option.

**Does this PR introduce a user interface change?**: For superusers, adds an entry in the 'Edit'/'General Information' form for a dataverse to change the store for a given dataverse.

**Is there a release notes update needed for this change?**: Yes - existing installations will have to update some jvm options to move to this version. The config docs describe the changes needed for a file store or s3 store.

**Additional documentation**:
DCM/rsync has hardcoded using a vanilla s3 store. Since I don't have a good way to test any rsync changes, I didn't try to update it to support multiple stores, but did try to make changes to enable an existing instance to keep using DCM/rsync. Adding the jvm options to define an s3 store with id 's3', as described in the docs, should be enough to make DCM work with this update. It should be possible to make the rsync capability available per store, and support stores with specified endpoints, etc. in the future.

carlosmcgregor commented on src/main/java/edu/harvard/iq/dataverse/dataaccess/SwiftAccessIO.java:
```diff
@@ -48,43 +48,48 @@
     //TODO: should this be dynamically generated based on size of file?
     //Also, this is in seconds
-    private static int TEMP_URL_EXPIRES = Integer.parseInt(System.getProperty("dataverse.files.swift.temporaryUrlExpiryTime", "60"));
+    private int tempUrlExpires = 60;
```
Why has this been hardcoded? Is it not compatible w/ the new driver ID configuration format?

qqmyers replied:
The static value has been replaced by one specific to the swift store. It is set to a default of 60 and then in line 73 above is reset based on the property. (Nominally it's redundant to set it here and then set a default on line 73 as well...)

carlosmcgregor commented on src/main/java/edu/harvard/iq/dataverse/dataaccess/SwiftAccessIO.java:
```diff
@@ -48,43 +48,48 @@
-    private String swiftFolderPathSeparator = System.getProperty("dataverse.files.swift.folderPathSeparator", "_");
-    private String swiftDefaultEndpoint = System.getProperty("dataverse.files.swift.defaultEndpoint");
+    private boolean isPublicContainer = true;
+    private String swiftFolderPathSeparator = "_";
```
This has also been hardcoded. @poikilotherm can probably answer if this is something needed by the community. I don't mind this being hard coded in general.

carlosmcgregor replied:
This has also been hardcoded. @poikilotherm can probably answer if this is something needed by the community. I don't mind this being hard coded in general.

pdurbin commented:
> I remember us having a conversation re: `-` in the configuration file.

@carlosmcgregor sorry, can you please remind me if you wanted a different character?

carlosmcgregor commented:
Quoting @poikilotherm from [this comment](https://github.com/IQSS/dataverse/pull/5680#issuecomment-475868350): 

> At least pretty please avoid `-` and `_` in the config names. `-` is an invalid char for environment variables, `_` is used for substitution. See IQSS/dataverse-kubernetes#11 Best option is to use a "." for namespacing as used elsewhere plus camel / pascal case.

pdurbin commented:
> At least pretty please avoid `-` and `_` in the config names.

Ok. Thanks, @carlosmcgregor . @poikilotherm do you still feel this way?

landreev commented:
@qqmyers  This is looking very good!
A couple of questions/requests:

- Is there a use case for having the storage driver selection menu on the dataverse page? - Somebody asked during our tech hour today if making it configurable via API only would be enough. 

- I very much like that your solution makes it possible to keep any already existing storage identifiers unchanged. But it would help to include an explicit migration instruction; something we could include in the release notes. It was very easy for me to figure out what settings needed to be changed/added for our existing S3 buckets here, by reading your config guide section.  But for other installations deploying  the release, we should at least warn them upfront that their S3 and/or swift storage volumes are going to stop working unless reconfigured.  Something like `doc/release-notes/6485-multiple-storage.md` using our normal notation. 

Thank you!

qqmyers commented:
@landreev - Thanks. The question as to whether the ability to configure a dataverse needs a GUI option is probably best answered by TDL (@CCMumma et. al.). With their instance supporting multiple universities, with local admins, having some way for the local admins to allow users to access a particular store (which may require approval for large data, or a cost, etc.) is useful and probably important long-term. That said, if it turns out to be rare, an API option may be enough for now (and it's not clear if associating a store with a dataverse is the only/best way to manage this longer term). I will note that it only shows for superusers right now, so the dataverse page is unchanged unless you have the privilege to make the change.

W.r.t. release notes - I agree that release notes are in order. I haven't kept up to date with where they should go before the final notes come out, but with your hint above, I'll go ahead and move the backward compatibility info from the docs (as @pdurbin wanted) and into some release notes.

pdurbin commented:
@qqmyers thanks. I'd feel remiss if I didn't mention that there are now some merge conflicts so if you would please resolve them at your convenience, it would be most appreciated! I assume they're from pull request #6517.

landreev commented:
@qqmyers 
> I will note that it only shows for superusers right now

Yes, we saw that. And, to be clear, I personally don't have a problem with that selection being there. I believe our UI designer asked the question because our normal approach was to keep superuser-only functionality out of the main pages. But, there are exceptions of course, and if there is a use case for it, it should be ok, IMO.

scolapasta commented:
Also, can this logic use the "root" logic that we do for other dataverse level configurations, like facets and metadatablocks? i.e. a dataverse should inherit this from its parent unless its overridden to be a new root.

(I'm assuming this makes sense as the main use case - if I create a child dataverse in a dataverse, it should be default use what's in its parent, rather than it's own.

Also, what happens if you switch it? do some files then live on the old store, while others on the new store?

landreev commented:
> @gudrand: Also, what happens if you switch it? do some files then live on the old store, while others on the new store?

Yes, similar to how it works now, if you change the dataverse-wide storage driver default - any pre-existing files stay where they are (and continue to be accessible); any new files get saved in the new storage volume.

qqmyers commented:
Summary response:
- will update and merge soon
- for the UI - rather than questioning having the ability to switch stores in the UI, it might be worth asking whether it would be better to allow dataverse admins or other role to manage it rather than superuser. That's an open question for TDL w.r.t. how they want to use this. I wasn't sure, so I just picked something very restrictive to start.
-for inheritance - my guess is that inheritance makes sense. Not sure if it's something to add in this PR or later, or if it needs to be optional like inheriting roles.
- for existing files - works as @landreev said. Overall, some tool/API call to migrate files between stores might be useful. Might have helped when there was a single store, but with multiple stores, it could become a more common thing to want to do. That said, it might be nice to nail down whether managing stores by dataverse is the right model longer term, or if we'll end up wanting something where files of a given size/type/sensitivity go to particular stores. That doesn't conflict with wanting to move files around, but might suggest that it needs to be a file/dataset level option than something you do for everything in a dataverse. (Also - fwiw: manual moves, as people had to do for going from files to S3 are still about the same degree of difficulty - you do the actual file transfers and then the storageidentifiers need to be updated to match (just using store ids instead of store types as the prefix)).

djbrooke commented:
Thanks @qqmyers, let us know when it's ready for us to take another look. If you have a chance, please drop in a screenshot of the switch placement so our design team can take a look. If not, we'll spin up the branch when it's back in code review. Thanks for the valuable contribution!

qqmyers commented:
@djbrooke - changed the docs to move upgrade info to release notes and merged with develop. So - I think that's all that's been requested so far/ready for more review.
Also - here's the UI change. The selection is from a pull down list of the defined store labels (not a type in).

![image](https://user-images.githubusercontent.com/6731983/72459117-0fed7b80-3798-11ea-84f1-a305484c3130.png). The default popup help text is 'A storage service to be used for datasets in this dataverse.'

scolapasta commented:
@qqmyers I'd say if inheritance is the right use case / model for this, we should probably have it as part of this PR.

qqmyers commented:
@scolapasta  - OK - just committed a chnage that I think will work for the 'Add Dataverse' option in the UI. As far as I can tell, the add Dataverse API call doesn't do inheritance, but I went ahead and made the storage driver something you can get/set with the json you get/send when you GET/POST a dataverse.

scolapasta commented:
@qqmyers this is a different approach than I was suggesting, and we can think about if it's enough. My idea was to mimic the other "roots" that there is a boolean (storageDriverRoot). If false, the getStorageDriver method would get the value of its parent. (in other words, if the parent changes, so does the child). If the boolean is true, then you're "detaching" that branch of dataverses and it would return whatever storage driver was directly associated with it.

For examples, see the facet root or the metadatablocks root.

qqmyers commented:
@scolapasta - ah, I see what you were talking about now. I don't have any big objection to doing it that way, but I see some minor differences in the use case, i.e. in that this is a single setting versus a whole set of options (so maybe not as much to be gained), and that, since existing files are in the original store, maybe automatically inheriting down the tree would be less useful. Does the fact that this has a global default change things too? You really have the options to use the default (dynamically change when it does), inherit from parent (and change when it does) or hardcode one of the choices. (Might at least change the UI where you could still have a list with 'default' and 'inherit' options in addition to the labels for the defined stores, instead of a separate checkbox/graying out the list).

djbrooke commented:
Thanks @qqmyers. I made one text change based on some design feedback.

djbrooke commented:
@scolapasta @qqmyers - is this ready for a quick review and then QA or does it need more discussion? Let me know if I need to offer opinions on anything.

qqmyers commented:
After discussion in Norway, I'm going to go ahead and alter the inheritance mechanism to align with the idea of having a root - recursing to find the first parent for which the store is defined. NO changes to the main multistore mechanism expected, just to the mechanism to select a store in a dataverse.

djbrooke commented:
@qqmyers nice. Back to code review for this, or are you planning on making additional changes?

qqmyers commented:
@djbrooke - update complete. The selections for the store assigned to a dataverse now include getting whichever store is the inherited default, whether that's the system default or inherited from an ancestor dataverse:

![image](https://user-images.githubusercontent.com/6731983/73573164-b4cab280-4440-11ea-9a7d-04d4e4895a0e.png)

qqmyers commented:
@poikilotherm - if you have questions when reviewing, let me know and I can pop into the chat room.

djbrooke commented:
Thanks @qqmyers. I may make a few doc edits early next week but this can for sure start code review. Thanks again and have a good weekend!

kcondon commented:
@poikilotherm Are you finished reviewing this?

kcondon commented:
@qqmyers Can you update this branch from develop? There are conflicts.

pdurbin commented:
I just noticed a merge conflict with IngestServiceBean.java. @qqmyers do you mind merging the latest from develop?

poikilotherm commented:
I haven't had a chance to review yet due to moving to production... I don't want to block someone, so please feel free to go ahead.

qqmyers commented:
@kcondon, @pdurbin  - this and #6490 merged with the latest develop branch

kcondon commented:
@qqmyers I followed the instruction in the upgrade release notes but when I add a dataset and upload the file I get an exception with the following server log entry:
```
[2020-02-05T18:11:15.056-0500] [glassfish 4.1] [SEVERE] [] [] [tid: _ThreadID=53 _ThreadName=Thread-9] [timeMillis: 1580944275056] [levelValue: 1000] [[
  java.io.IOException: getDataAccessObject: Unsupported storage method.
	at edu.harvard.iq.dataverse.dataaccess.DataAccess.getStorageIO(DataAccess.java:88)
	at edu.harvard.iq.dataverse.search.IndexServiceBean.addOrUpdateDataset(IndexServiceBean.java:978)
	at edu.harvard.iq.dataverse.search.IndexServiceBean.addOrUpdateDataset(IndexServiceBean.java:689)
	at edu.harvard.iq.dataverse.search.IndexServiceBean.indexDataset(IndexServiceBean.java:448)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.glassfish.ejb.security.application.EJBSecurityManager.runMethod(EJBSecurityManager.java:1081)
	at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:1153)
	at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4786)
	at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:656)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
	at org.jboss.weld.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:64)
	at org.jboss.weld.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
	at sun.reflect.GeneratedMethodAccessor115.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:163)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:140)
	at sun.reflect.GeneratedMethodAccessor114.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:369)
	at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4758)
	at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4746)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:212)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:88)
	at com.sun.proxy.$Proxy310.indexDataset(Unknown Source)
	at edu.harvard.iq.dataverse.search.__EJB31_Generated__IndexServiceBean__Intf____Bean__.indexDataset(Unknown Source)
	at edu.harvard.iq.dataverse.engine.command.impl.UpdateDatasetVersionCommand.onSuccess(UpdateDatasetVersionCommand.java:227)
	at edu.harvard.iq.dataverse.EjbDataverseEngine.completeCommand(EjbDataverseEngine.java:329)
	at edu.harvard.iq.dataverse.EjbDataverseEngine.submit(EjbDataverseEngine.java:263)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.glassfish.ejb.security.application.EJBSecurityManager.runMethod(EJBSecurityManager.java:1081)
	at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:1153)
	at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4786)
	at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:656)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
	at org.jboss.weld.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:64)
	at org.jboss.weld.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
	at sun.reflect.GeneratedMethodAccessor115.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:163)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:140)
	at sun.reflect.GeneratedMethodAccessor114.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:369)
	at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4758)
	at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4746)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:212)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:88)
	at com.sun.proxy.$Proxy281.submit(Unknown Source)
	at edu.harvard.iq.dataverse.__EJB31_Generated__EjbDataverseEngine__Intf____Bean__.submit(Unknown Source)
	at edu.harvard.iq.dataverse.DatasetPage.save(DatasetPage.java:3421)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.sun.el.parser.AstValue.invoke(AstValue.java:289)
	at com.sun.el.MethodExpressionImpl.invoke(MethodExpressionImpl.java:304)
	at org.jboss.weld.util.el.ForwardingMethodExpression.invoke(ForwardingMethodExpression.java:40)
	at org.jboss.weld.el.WeldMethodExpression.invoke(WeldMethodExpression.java:50)
	at com.sun.faces.facelets.el.TagMethodExpression.invoke(TagMethodExpression.java:105)
	at javax.faces.component.MethodBindingMethodExpressionAdapter.invoke(MethodBindingMethodExpressionAdapter.java:87)
	at com.sun.faces.application.ActionListenerImpl.processAction(ActionListenerImpl.java:102)
	at javax.faces.component.UICommand.broadcast(UICommand.java:315)
	at javax.faces.component.UIViewRoot.broadcastEvents(UIViewRoot.java:790)
	at javax.faces.component.UIViewRoot.processApplication(UIViewRoot.java:1282)
	at com.sun.faces.lifecycle.InvokeApplicationPhase.execute(InvokeApplicationPhase.java:81)
	at com.sun.faces.lifecycle.Phase.doPhase(Phase.java:101)
	at com.sun.faces.lifecycle.LifecycleImpl.execute(LifecycleImpl.java:198)
	at javax.faces.webapp.FacesServlet.service(FacesServlet.java:646)
	at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1682)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:344)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
	at org.glassfish.tyrus.servlet.TyrusServletFilter.doFilter(TyrusServletFilter.java:295)
	at org.apache.catalina.core.ApplicationFi]]

[2020-02-05T18:11:15.056-0500] [glassfish 4.1] [SEVERE] [] [] [tid: _ThreadID=53 _ThreadName=Thread-9] [timeMillis: 1580944275056] [levelValue: 1000] [[
  lterChain.internalDoFilter(ApplicationFilterChain.java:256)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
	at org.ocpsoft.rewrite.servlet.RewriteFilter.doFilter(RewriteFilter.java:226)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:256)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:316)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:160)
	at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:734)
	at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:673)
	at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:99)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:174)
	at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:734)
	at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:673)
	at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:412)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:282)
	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:459)
	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:167)
	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:201)
	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:175)
	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:235)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:284)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:201)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:133)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:112)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:561)
	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:112)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:117)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:56)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:137)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:565)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:545)
	at java.lang.Thread.run(Thread.java:748)]]

[2020-02-05T18:11:16.052-0500] [glassfish 4.1] [WARNING] [] [edu.harvard.iq.dataverse.dataset.DatasetUtil] [tid: _ThreadID=52 _ThreadName=jk-connector(1)] [timeMillis: 1580944276052] [levelValue: 900] [[
  getThumbnail(): Failed to initialize dataset StorageIO for null://doi:10.70122/FK2/PR2QMC (getDataAccessObject: Unsupported storage method.)]]
``

qqmyers commented:
@kcondon Can you share the bin/asadmin list-jvm-options results (minus any passwords?)

kcondon commented:
@qqmyers 
```
./asadmin list-jvm-options
-server
-Djava.awt.headless=true
-Djdk.corba.allowOutputStreamSubclass=true
-Djavax.xml.accessExternalSchema=all
-Djavax.management.builder.initial=com.sun.enterprise.v3.admin.AppServerMBeanServerBuilder
-XX:+UnlockDiagnosticVMOptions
-Djava.endorsed.dirs=${com.sun.aas.installRoot}/modules/endorsed${path.separator}${com.sun.aas.installRoot}/lib/endorsed
-Djava.security.policy=${com.sun.aas.instanceRoot}/config/server.policy
-Djava.security.auth.login.config=${com.sun.aas.instanceRoot}/config/login.conf
-Dcom.sun.enterprise.security.httpsOutboundKeyAlias=hmdc-wildcard
-Djavax.net.ssl.keyStore=${com.sun.aas.instanceRoot}/config/keystore.jks
-Djavax.net.ssl.trustStore=${com.sun.aas.instanceRoot}/config/cacerts.jks
-Djava.ext.dirs=${com.sun.aas.javaRoot}/lib/ext${path.separator}${com.sun.aas.javaRoot}/jre/lib/ext${path.separator}${com.sun.aas.instanceRoot}/lib/ext
-Djdbc.drivers=org.apache.derby.jdbc.ClientDriver
-DANTLR_USE_DIRECT_CLASS_LOADING=true
-Dcom.sun.enterprise.config.config_environment_factory_class=com.sun.enterprise.config.serverbeans.AppserverConfigEnvironmentFactory
-Dorg.glassfish.additionalOSGiBundlesToStart=org.apache.felix.shell,org.apache.felix.gogo.runtime,org.apache.felix.gogo.shell,org.apache.felix.gogo.command,org.apache.felix.shell.remote,org.apache.felix.fileinstall
-Dosgi.shell.telnet.port=6666
-Dosgi.shell.telnet.maxconn=1
-Dosgi.shell.telnet.ip=127.0.0.1
-Dgosh.args=--nointeractive
-Dfelix.fileinstall.dir=${com.sun.aas.installRoot}/modules/autostart/
-Dfelix.fileinstall.poll=5000
-Dfelix.fileinstall.log.level=2
-Dfelix.fileinstall.bundles.new.start=true
-Dfelix.fileinstall.bundles.startTransient=true
-Dfelix.fileinstall.disableConfigSave=false
-XX:NewRatio=2
-Dcom.ctc.wstx.returnNullForDefaultNamespace=true
-XX:MaxPermSize=512m
-XX:PermSize=256m
-XX:+UseG1GC
-Xmx5981m
-Ddataverse.files.directory=/usr/local/glassfish4/glassfish/domains/domain1/files
-Ddataverse.rserve.host=localhost
-Ddataverse.rserve.port=6311
-Ddataverse.rserve.user=xxxxxxxx
-Ddataverse.rserve.password=xxxxxxxx
-Ddataverse.fqdn=dvn-build.hmdc.harvard.edu
-Ddataverse.siteUrl=https://${dataverse.fqdn}
-Ddataverse.auth.password-reset-timeout-in-minutes=60
-Djavax.xml.parsers.SAXParserFactory=com.sun.org.apache.xerces.internal.jaxp.SAXParserFactoryImpl
-Ddoi.password=xxxxxx
-Ddoi.username=xxxxxx
-Ddataverse.dropbox.key=xxxxxxxx
-Ddataverse.timerServer=true
-Ddataverse.handlenet.admcredfile=xxxxxxx
-Ddataverse.files.storage-driver-id=s3
-Ddoi.mdcbaseurlstring=https://api.test.datacite.org
-Ddoi.baseurlstring=https://mds.test.datacite.org
-Ddataverse.files.s3.type=s3
-Ddataverse.files.s3.label=s3
-Ddataverse.files.s3.bucket-name=xxxxxx
Command list-jvm-options executed successfully.
```

qqmyers commented:
@kcondon - so far I don't see any issue with the config/documentation and think these are probably real things to fix. TDL isn't running full-text indexing so testing there wouldn't have hit this code, and the thumbnail issue in the log might not have been hit either (or may be non-fatal so only viewable in the log). In any case, it will probably take me a little time to dig into it.

qqmyers commented:
@kcondon - this should be ready to test again. The issue was caused by the update to the inheritance mechanism and affected the assignment of a store to dataset entries in the dvobject table which then affected full text indexing and thumbnails. More specifically, when the storage driver wasn't directly set on the enclosing dataverse, the driver was recorded as 'null' and the fix was to recurse upward through parents until the effective store was found.

Aside from the software fix, if you are testing from an instance with database entries, the any datasets will still have the problem - you should be able to delete them or change their storageidentifiers to start with the right store instead of 'null://' in the dvobject table. New datasets shouldn't be affected. 

FWIW: I've confirmed that new datasets are created correctly/right value in the dvobject table, but did not specifically test full-text indexing.

kcondon commented:
@qqmyers Still getting exception on full text indexing.

qqmyers commented:
@kcondon - how are you creating the file? web interface upload or API?

kcondon commented:
@qqmyers Just uploading it through the UI

qqmyers commented:
@kcondon - one more question - uploading the file when you create the dataset or after it exists?

When I turn on full-text indexing at TDL, I don't see this so there are two possibilities: some config error or a difference between their branch and the PR - either because I missed a commit or due to some merge issue in keeping the PR up to date. The first couple of lines suggest that the problem is during ingest rather than indexing - indexing just reuses the same bad value.

I guess the best thing for me to do is to use my new GDCC credentials from Don to spin up a test instance of this PR. So - back to me I guess.

kcondon commented:
@qqmyers For this particular test the file was added on create but in earlier iterations I tested:
1. on create
2. after create, prepublish
3. while creating a draft of a published version by uploading a file.

I'm told each of those has a unique path and so in the past I use them to test behavior.

qqmyers commented:
I have a clue now - looks like there is some code in handling temp files that depends on a default 'file' store existing (through some if store=null, store='file' logic somewhere).

pdurbin commented:
> I guess the best thing for me to do is to use my new GDCC credentials from Don to spin up a test instance of this PR. So - back to me I guess.

@qqmyers I've created AWS/EC2 accounts for @donsizemore @poikilotherm and @4tikhonov and I'm happy to create one for you too. Please just let me know! :smile:

qqmyers commented:
@kcondon - I think this is fixed (again) but I have yet to test it. Feel free to give it a go or wait.

The issue relates to the fact that temporary files are initially created with storageidentifiers that have no driver prepended (e.g. they don't have file:// or s3://). The original code made the assumption in that case that the store type was 'file' (which was true for all temp files). With the multistore change, assigning a default store of 'file' indicates a store with that id, not the file type. So, if that store is not defined, the type was 'Undefined', which is not supported (not 'file','s3', or 'swift'). TDL has such a store (since I was testing having file and s3 stores at the same time), so it accidentally worked for temp files as well. Your test instance presumably didn't define a file store and therefore failed - skipping the code that should have moved the file from temp storage to the driver and updated the storageidentifier with the correct prefix. If I understand correctly, that wasn't fatal but left the storageidentifier with no prefix which then caused failures in any code that tries to retrieve the file.

So - I've redefined the default storageId assigned for case where storageidentifier doesn't include one as "", and changed the logic that finds the store type from the id to have a special case if the storeId is "" or "tmp" - those are now of type "tmp" and the IngestServiceBean.saveAndAddFilesToDataset() method now checks for that type only.

Which should work.

(FWIW: There are a number of places in the code where a storageidentifier is assumed to be a temp object if it starts with "tmp://" or has no "*://" prefix at all. Some of those appear to be broken for the "tmp://" prefix case, so I suspect that this is now a legacy/deprecated/no longer used case (temp dvobjects never have a prefix), but since I can't prove that, I've tried to make this work either way. Someday it would be nice to push more of the logic creating/interpreting storageidentifiers into the StoreIO classes...)

qqmyers commented:
@kcondon - looks like the fix works at TDL - back to you.

kcondon commented:
@qqmyers That works, I'll finish testing, thanks.

landreev commented on scripts/installer/glassfish-setup.sh:
```diff
@@ -73,6 +73,9 @@ function preliminary_setup()
   # location of the datafiles directory: 
   # (defaults to dataverse/files in the users home directory)
   ./asadmin $ASADMIN_OPTS create-jvm-options "\-Ddataverse.files.directory=${FILES_DIR}"
```
@qqmyers After some testing and looking into this with @kcondon, we are wondering - since it is now possible to configure more than one stores of the type "file", it seems like we need to be able to also define different physical directories for these locations. 
i.e., instead of the fixed `Ddataverse.files.directory=...`, it should be
`dataverse.files.file.directory=...`, `dataverse.files.file1.directory=...`, etc.
(but maybe we should also leave the current `dataverse.files.directory`, for backward compatiblity, to be consulted if `dataverse.files.file.directory` is not defined - ?)
what do you think?

landreev replied:
@qqmyers OK, please [half]scratch the above - I see what's going on... the code is already doing what I suggested above; it's the documentation and the migration instruction need to be updated, to include the .directory JVM option. (otherwise the file storage is working, BUT the files end up stored in /tmp/files!)

qqmyers replied:
Yep - was just writing that up since I knew I'd changed the code. I can update the docs to match.

landreev commented:
@qqmyers Also, as heads up, next time you sync up with the develop branch, you'll add my most recent flyway script to your branch; which is numbered `V4.19.0.0__...` - preceding the one in your branch (`V4.19.0.1__...`). So I believe you'll need to remove the `V4.19.0.1` entry from your `flyway_schema_history`table, or else your branch will fail to deploy with a flyway error. 

Sorry for the extra trouble - this felt like the least messy way of resolving it.

qqmyers commented:
@landreev - OK - I've updated the docs, added setting the dir for the 'file' store in glassfish-setup.sh and changed the sql script name to include the issue # instead of tdb - I don't think the last messes with your change. If you guys think further doc changes are needed, feel free to edit directly.

kcondon commented:
@qqmyers I've made slight updates to the doc, thanks. However, have found an issue with 
```
-Ddataverse.files.storage-driver-id=<id>
```
It works fine when id=file but fails with id=file1.
When I create a dataset without uploading a file, I see error in log and similar error when uploading a file:

```
[2020-02-11T13:01:05.157-0500] [glassfish 4.1] [INFO] [] [edu.harvard.iq.dataverse.engine.command.impl.AbstractCreateDatasetCommand] [tid: _ThreadID=53 _ThreadName=jk-connector(2)] [timeMillis: 1581444065157] [levelValue: 800] [[
  Failed to create StorageIO. StorageIdentifier set to default. Not fatal.(createDataAccessObject: Unsupported storage method file1)]]

[2020-02-11T13:01:07.470-0500] [glassfish 4.1] [WARNING] [] [edu.harvard.iq.dataverse.dataset.DatasetUtil] [tid: _ThreadID=52 _ThreadName=jk-connector(1)] [timeMillis: 1581444067470] [levelValue: 900] [[
  getThumbnail(): Failed to initialize dataset StorageIO for file1://doi:10.70122/FK2/AQKIQM (getDataAccessObject: Unsupported storage method.)]]

[2020-02-11T13:01:07.471-0500] [glassfish 4.1] [WARNING] [] [edu.harvard.iq.dataverse.dataset.DatasetUtil] [tid: _ThreadID=52 _ThreadName=jk-connector(1)] [timeMillis: 1581444067471] [levelValue: 900] [[
  getThumbnail(): Failed to initialize dataset StorageIO for file1://doi:10.70122/FK2/AQKIQM]]

```

qqmyers commented:
@kcondon - could be code, but to verify, you have
./asadmin create-jvm-options "\-Ddataverse.files.file.type=file"
    ./asadmin create-jvm-options "\-Ddataverse.files.file1.label=file1"
    ./asadmin create-jvm-options "\-Ddataverse.files.file1.directory=<your directory>"

(first entry is still type file, others have the file1 id)

kcondon commented:
@qqmyers I do not have what you described since one item appears incorrect:
./asadmin create-jvm-options "-Ddataverse.files.file.type=file" should be
./asadmin create-jvm-options "-Ddataverse.files.file1.type=file"
```
        <jvm-options>-Ddataverse.files.storage-driver-id=file1</jvm-options>
        <jvm-options>-Ddataverse.files.file1.directory=/usr/local/glassfish4/glassfish/domains/domain1/files1</jvm-options>
        <jvm-options>-Ddataverse.files.file1.type=file</jvm-options>
        <jvm-options>-Ddataverse.files.file1.label=file</jvm-options>
```

qqmyers commented:
@kcondon - any chance there's a second jvm-option that has the type as file1? OR a lack of restart after an update to the jvm-options? 
So far when I trace through the code, I quickly get to a 
System.getProperty("dataverse.files." + driverId + ".type", "Undefined"); 
call whose result should be what shows up in the 'createDataAccessObject: Unsupported storage method file1' message.

kcondon commented:
@qqmyers No, I had double checked the restart, good. Here is the result of grepping on file1:
```
grep file1 domain.xml
        <jvm-options>-Ddataverse.files.storage-driver-id=file1</jvm-options>
        <jvm-options>-Ddataverse.files.file1.directory=/usr/local/glassfish4/glassfish/domains/domain1/files1</jvm-options>
        <jvm-options>-Ddataverse.files.file1.type=file</jvm-options>
        <jvm-options>-Ddataverse.files.file1.label=file</jvm-options>
```

landreev commented:
I don't know if it helps, but looking at the code in `DataAccess.populateDrivers()`, I'm seeing this: 
```
   logger.info("Found Storage Driver: " + driverId + " for " + p.get(property).toString());
```
yet I'm not seeing the message above in the server logs on the system where @kcondon has been testing the branch.

qqmyers commented:
@kcondon, @landreev - I've made some changes that should fix things. It looks like some/all ways of creating file store entries left the storageidentifier without a \<diverId\>:// prefix, unlike the s3 and swift stores, so things only worked sometimes or if one used the same dir for the default and the file1 store, etc. What that involves is updating the db script so that all existing storageidentifiers for files that have no prefix get a file:// one (consistent with having to id the default store 'file' in the config options), and a change to the FileStorageIO class to assign the driver prefix upon write (not sure that's the best place to do it, but it's where the s3 and swift stores do it). There was also a place in delete where the text string 'file' was being prepended instead of using the driverId.

With those, I think the test cases you had above should work.

I also realized that the datasets were getting an extra 'doi:' in their storageidentifiers from one path, so I updated that, and also found a couple places where things may not have been dealing with alt identifiers correctly.

W.r.t. not seeing the 'Found' driver statements, those are only populated/cached when someone want's to set the storage in the Edit/General Info Dataverse page. So I don't think that not seeing them was related. (That log could/should be .fine()  at some point, unless documenting the set of storage drivers available once per startup makes sense.)

Hopefully good to go again. I've been trying to use multiple s3 and file stores on TDL's dev server and so far so good.

djbrooke commented:
Thanks @qqmyers, I'll move this back over to QA.

kcondon commented:
@qqmyers Thanks for all the work! I'm testing this now. Btw, Jenkins reports the build as unstable due to tests failing. This is from the build console:
```
[ERROR] Failures: 
[ERROR]   FileAccessIOTest.testGetStorageLocation:214 expected:<[file]:///tmp/files/tmp/da...> but was:<[dummmy]:///tmp/files/tmp/da...>
[INFO] 
[ERROR] Tests run: 1121, Failures: 1, Errors: 0, Skipped: 6
[INFO] 
[ERROR] There are test failures.
```

qqmyers commented:
@kcondon - test is fixed

kcondon commented:
@qqmyers Thanks for the work, was able to configure and use 2 s3 profiles, 2 local file stores. Did notice a warning when using local file store and publishing a dataset, which triggers export:

[2020-02-18T19:11:51.930-0500] [glassfish 4.1] [WARNING] [] [edu.harvard.iq.dataverse.engine.command.impl.FinalizeDatasetPublicationCommand] [tid: _ThreadID=52 _ThreadName=jk-connector(2)] [timeMillis: 1582071111930] [levelValue: 900] [[
  Dataset publication finalization: exception while exporting:IO Exception thrown exporting as export_dcterms.cached]]

I did not see export_dcterms.cached file on the file system but that did not appear to cause and problems. Any thoughts?

qqmyers commented:
@kcondon - and thanks to you for great testing!

For the exporter error, it is hard to see how the changes here could affect one export and not others, but if dcterms is the first one (you're not seeing any cached exports), it could be related to multi-stores (the export code does a ~non-standard thing - it creates a fake 'placeholder' storageIO instance that it only uses to write the aux file - using two different methods in two different cases).

FWIW - I'm not sure the ExportService is doing what's expected either: if the code to create an outputStream from the storageIO works on line https://github.com/IQSS/dataverse/blame/develop/src/main/java/edu/harvard/iq/dataverse/export/ExportService.java#L258, it looks like tempFileRequired is false and the code at https://github.com/IQSS/dataverse/blame/develop/src/main/java/edu/harvard/iq/dataverse/export/ExportService.java#L269 should run, which stores to a local temp file only. It looks like an IOException has to occur in the block with the first line so that code at https://github.com/IQSS/dataverse/blame/develop/src/main/java/edu/harvard/iq/dataverse/export/ExportService.java#L275 runs instead. 

If that analysis is correct (if you concur I can raise an issue), it would mean that proper functioning has required failure of the initial attempt to stream the aux file, raising the odd possibility that a fix in the multistore code might be exposing an error here. That said, it still seems like this would be an all-formats or none issue.

pdurbin commented:
I'm catching up but my take on the last comment is that we're using #6666 to track the issue with exports.
